### PR TITLE
add retroImplements to lf/iface DefInterface

### PR DESCRIPTION
**Pull request by S11001001**
Add retroactive interface-implements-template resolution, so this is [more or less] in place when the LF/language feature lands.

Part of #14081. Excludes serious testing that can be added when the whole feature can be tested properly.

* [x] variant operating on interface sets, for Java codegen & json-api
* [x] empty changelog

### CODE REVIEW

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala [hunk 1]
```diff
@@ -62,6 +62,17 @@ private class PackageService(
       val findIface = iface.Interface.findAstInterface(Function unlift lookupIf)
       diff.transform((_, iface) => iface resolveChoicesAndIgnoreUnresolvedChoices findIface)
     }
+
+    private[this] def appendAndResolveRetroactiveInterfaces(diff: PackageStore): PackageStore = {
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala:66]: For new readers the `PackageStore` type alias is just another thing to lookup in the code.  Using `Map[String, Interface]` makes that lookup unnecessary.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala [hunk 2]
```diff
@@ -62,6 +62,17 @@ private class PackageService(
...
+      def lookupIf(packageStore: PackageStore, pkId: Ref.PackageId) =
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala:67]: So this probably returns `Option[(Interface, Interface => Map[String,Interface])]` and I guess that is used in a GetterSetterAt? (I think this refers to you telling me earlier you did not use lenses but just functions)

**Review by remyhaemmerle-da — COMMENTED** [review]
This looks reasonable to me.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala (hunk 1 above)
Comment (ray-roestenburg-da) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala:66]: I guess this is more an issue with me reading through these diffs than with a type alias..

**Review by ray-roestenburg-da — APPROVED** [review]
LGTM

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala [hunk 3]
```diff
@@ -202,6 +212,29 @@ final case class DefTemplate[+Ty](
...
+    import TemplateChoices.{Resolved, Unresolved}
+    copy(
+      implementedInterfaces = implementedInterfaces :+ ifaceName,
+      tChoices = tChoices match {
+        case unr @ Unresolved(_, sources) =>
+          unr.copy(unresolvedChoiceSources = sources incl ifaceName)
+        // If unresolved, we need only add ifc as a future interface to resolve;
+        // otherwise, we must self-resolve and add to preexisting resolutions
+        case r @ Resolved(rc) =>
+          type K[C] = Semigroup[Resolved.Choices[C]]
```
Comment (ray-roestenburg-da) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala:229]: Why is this needed?

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala [hunk 4]
```diff
@@ -202,6 +212,29 @@ final case class DefTemplate[+Ty](
...
+          r.copy(resolvedChoices =
+            FirstVal
+              .unsubst[K, TemplateChoice[OTy]](Semigroup.apply)
+              .append(rc, ifc choicesAsResolved ifaceName)
```
Comment (ray-roestenburg-da) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala:233]: I'm assuming the choices resolved with if and ifaceName are appended to rc, but I don't understand why the Firstval, semigroup stuff is needed.

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala (hunk 3 above)
Comment (S11001001) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala:229]: We have a sensible semigroup for almost the whole resolved choices structure, but we need a way to "combine" single choices; the rule we choose is "just use the first one". We use `K` to specify that `FirstVal` is our chosen rule.

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala (hunk 4 above)
Comment (S11001001) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala:233]: The semigroup is the same one used for aggregating interface choices [when doing choice resolution](https://github.com/digital-asset/daml/blob/8d21d19133a1ca507760fb1b875ece3d6ad2d6db/daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala#L287-L301). So we're just doing one step of that aggregation here.
